components/match2/commons/opponent.lua

```diff
@@ -0,0 +1,331 @@
+---
+-- @Liquipedia
+-- wiki=commons
+-- page=Module:Opponent
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Array = require('Module:Array')
+local Flags = require('Module:Flags')
+local String = require('Module:StringUtils')
+local Table = require('Module:Table')
+local TeamTemplate = require('Module:TeamTemplate')
+
+--[[
+Structural type representation of an opponent.
+
+Examples:
+{type = 'solo', players = {displayName = 'Neeb'}}
+{type = 'team', template = 'alpha x 2020'}
+{type = 'literal', name = 'B2'}
+
+See Opponent.isOpponent for the exact encoding scheme and required fields.
+
+- For opponent display components, see Module:OpponentDisplay.
+- For input from wikicode, use {{1Opponent|...}}, {{TeamOpponent|...}} etc.
+- Used by: PrizePool, GroupTableLeague, match2 Matchlist/Bracket,
+StarcraftMatchSummary
+- Wikis may add additional wiki-specific fields to the opponent representation.
+
+]]
+local Opponent = {}
+
+Opponent.partyTypes = {'solo', 'duo', 'trio', 'quad'}
+Opponent.types = Array.extend(Opponent.partyTypes, {'team', 'literal'})
+
+Opponent.partySizes = {
+	solo = 1,
+	duo = 2,
+	trio = 3,
+	quad = 4,
+}
+
+function Opponent.typeIsParty(type)
+	return Opponent.partySizes[type] ~= nil
+end
+
+--[[
+Returns the player count for a party type, or nil otherwise.
+
+Opponent.partySize('duo') == 2
+]]
+function Opponent.partySize(type)
+	return Opponent.partySizes[type]
+end
+
+--[[
+Creates a blank opponent of the specified type
+]]
+function Opponent.blank(type)
+	if type == 'team' then
```

Comment:
```suggestion
	if type == TYPE_TEAM then
```

etc for all type usages in this file. Add these to Opponent like so:

```
Opponent.TYPE_TEAM = 0
```

Then use them from outside this module.

When doing so please also add an explainer comment for the term `party` above the constant for its type.

Reply:
That's not how the structural type is supposed to work. If I print an opponent in the debugger then I expect to see
```
{type = 'team', template = 'alpha x 2020'}
```

and not

```
{type = 0, template = 'alpha x 2020'}
```

The opponent types were chosen to match the allowed values in the `match2opponents.type` field. I think it's an important part of the design of this module.

Reply:
Make the constants strings then

Reply:
You mean like `Opponent.TYPE_TEAM = 'team'`? That doesn't help - if you look at the lpdb record, or print the object in the lua debugger, it shows up as `'team'` and not `'Opponent.TYPE_TEAM'`.

Plus `Opponent.TYPE_TEAM` is unnecessary verbose, and I don't want to set a bad example for others using `Module:Opponent`.

Reply:
Why doesn't it help? It helps me. I explained why in the first comment. I insist.

Reply:
It doesn't help solve the problem of there being two different ways of calling one thing, which is needlessly confusing. In lua you have `Opponent.TYPE_TEAM`, but everywhere else it is just `team`, and everybody already thinks of it as `team`. For example if I were to explain how `Opponent` works in discord, I would write `{type = 'team', template = 'alpha x'}`. Writing `{type = Opponent.TYPE_TEAM, template = 'alpha x'}` is just ridiculous. Also think of how this would appear in a template arg. The most natural is `{{PrizePool|type=team}}`. `{{PrizePool|type={{OpponentType|TEAM}}}}` is too verbose, and not inviting to editors.

In case you don't know, I spend quite a bit of effort on naming things properly and making modules easy to use by others. Removing constants like this is a simple way to make code cleaner and improve readability.

Reply:
OK I changed it to use `Opponent.team` which should close to you want you want. It looks a bit better than `Opponent.TYPE_TEAM`. I can also do `Opponent.type('team')` if you prefer that.

Reply:
This is okay